MapboxNavigation/SimulatedRoute.swift

```diff
@@ -5,85 +5,198 @@ protocol SimulatedRouteDelegate : class {
     func simulation(_ locationManager: CLLocationManager, didUpdateLocations locations: [CLLocation])
 }
 
+class SimulatedLocation: CLLocation {
+    var turnPenalty: Double = 0
+}
+
 class SimulatedRoute : NSObject {
-    
+    let maximumSpeed: CLLocationSpeed = 30 // ~108 kmh
+    let minimumSpeed: CLLocationSpeed = 5 // ~18 kmh
     var speed: CLLocationSpeed = 30
-    
     var distanceFilter: CLLocationDistance = 10
-    
     var verticalAccuracy: CLLocationAccuracy = 40
-    
     var horizontalAccuracy: CLLocationAccuracy = 40
+    var polyline = [CLLocationCoordinate2D]()
     
-    var locations = [CLLocation]()
+    // minimumSpeed will be used when a location have maximumTurnPenalty
+    fileprivate let maximumTurnPenalty: CLLocationDirection = 90
+    // maximumSpeed will be used when a location have minimumTurnPenalty
+    fileprivate let minimumTurnPenalty: CLLocationDirection = 0
     
+    var processedLocations: [CLLocation]!
     let locationManager = CLLocationManager()
-    
     weak var delegate: SimulatedRouteDelegate?
     
-    convenience init(_ polyline: [CLLocationCoordinate2D]) {
+    convenience init?(along polyline: [CLLocationCoordinate2D]) {
+        if polyline.count < 3 {
+            print("Not enough coordinates in polyline to simulate")
+            return nil
+        }
+        
         self.init()
+        self.polyline = polyline
+        
+        let turnPenaltyBasedLocations = calculateTurnPenalty(coordinates: polyline)
+        let interpolatedLocations = interpolate(locations: turnPenaltyBasedLocations)
+        let speedBasedLocations = calculateSpeed(for: interpolatedLocations)
+        
+        processedLocations = calculateTimestamp(for: speedBasedLocations)
+    }
+    
+    // Calculate timestamp when speed has been calculated.
+    fileprivate func calculateTimestamp(for locations: [SimulatedLocation]) -> [SimulatedLocation] {
```

Comment:
This method [has no side effects](https://swift.org/documentation/api-design-guidelines/#name-according-to-side-effects), so it should be named with a noun phrase instead of a verb. It doesn’t seem to use `self` in any way, so maybe it makes sense for it to be a standalone function, an extension method like `SimulatedLocation.simulatedLocationsBeginningNow(for:)`, or even a method extending `[SimulatedLocation]`.